### PR TITLE
**Breaking:** Rewrite Select + Add Custom Option + Make Accessible + Implement New Design

### PR DESCRIPTION
**Pull request by TejasQ**
# Summary
This PR:
- rewrites the `Select` component to use hooks and existing Operational components internally
- adds a custom input property to `Select`
- implements a new design

**Todo:**

- [x] move `Select` utilities out
- [x] cover with unit tests
- [x] cover with e2e tests
- [x] decide what to do about multiselect + custom option
- [x] Implement correct "selected" style

# Related Issue
[Jira](https://contiamo.atlassian.net/browse/UI-38)
[Screen](https://zpl.io/b647pYq)

# To be tested

Me
- [x] No error or warning in the console on `localhost:6060`

Tester 1

- [ ] Things look good on the demo.
  

Tester 2

- [ ] Things look good on the demo.

### CODE REVIEW

File: src/ContextMenu/ContextMenu.tsx [hunk 1]
```diff
@@ -6,6 +6,7 @@ import styled from "../utils/styled"
 
 import { keyCodes } from "../utils"
 import ContextMenuItem, { IContextMenuItem } from "./ContextMenu.Item"
+import noop = require("lodash/noop")
```
Comment (fabien0102) [inline src/ContextMenu/ContextMenu.tsx]: `require` ? 🤔

File: src/ContextMenu/ContextMenu.Item.tsx [hunk 2]
```diff
@@ -24,43 +25,58 @@ export interface IContextMenuItem<TValue = any> {
...
+    textAlign: align,
+    display: "flex",
+    alignItems: "center",
+    fontWeight: isActive ? theme.font.weight.bold : theme.font.weight.medium,
+    boxShadow: isActive ? activeShadow : "none",
+    ":focus": {
+      boxShadow: activeShadow,
+      outline: "none",
+    },
+    ...(Boolean(typeof item !== "string" && item.description)
```
Comment (stereobooster) [inline src/ContextMenu/ContextMenu.Item.tsx]:
```suggestion
    ...(typeof item !== "string" && Boolean(item.description)
```

File: src/ContextMenu/ContextMenu.Item.tsx [hunk 3]
```diff
@@ -24,43 +25,58 @@ export interface IContextMenuItem<TValue = any> {
...
+    backgroundColor: theme.color.white,
+    lineHeight: `${condensed ? 35 : 44}px`,
+    padding: `0 ${theme.space.content}px`,
+    textAlign: align,
+    display: "flex",
+    alignItems: "center",
+    fontWeight: isActive ? theme.font.weight.bold : theme.font.weight.medium,
+    boxShadow: isActive ? activeShadow : "none",
+    ":focus": {
+      boxShadow: activeShadow,
```
Comment (stereobooster) [inline src/ContextMenu/ContextMenu.Item.tsx]: we will have 3 focus shadows

File: src/ContextMenu/ContextMenu.Item.tsx [hunk 4]
```diff
@@ -24,43 +25,58 @@ export interface IContextMenuItem<TValue = any> {
...
-    : {
-        cursor: "not-allowed",
-        color: theme.color.text.lightest,
-      }),
-  borderTop: `1px solid ${theme.color.separators.default}`,
-  "&:last-child": {
-    paddingBottom: 2,
-  },
-}))
+const Container = styled("div")<Props>(({ align, theme, onClick, isActive, condensed, width, item }) => {
```
Comment (stereobooster) [inline src/ContextMenu/ContextMenu.Item.tsx]:
```suggestion
const ContextMenuItem = styled("div")<Props>(({ align, theme, onClick, isActive, condensed, width, item }) => {
```

based on `label: "contextmenuitem",`?

File: src/ContextMenu/ContextMenu.tsx (hunk 1 above)
Comment (stereobooster) [inline src/ContextMenu/ContextMenu.tsx]:
```suggestion
import noop from "lodash/noop";
```

**Comment by fabien0102** [comment]
Some feedbacks of points that are not on the mockup:

1. 
image

Can we add an ellipsis to the text on overflow?

2. 
image

The custom input is not really clear, as a dummy person I tried to click on `Custom…` 3 times before understand the component 😅

File: src/ContextMenu/ContextMenu.tsx [hunk 5]
```diff
@@ -151,29 +159,56 @@ class ContextMenu extends React.Component<ContextMenuProps, Readonly<State>> {
...
     const renderedChildren = isChildAFunction(children) ? children(this.state.isOpen) : children
     return (
       <>
         {isOpen && <InvisibleOverlay onClick={this.toggle} />}
-        <Container isOpen={isOpen} {...props} align={align} onClick={this.toggle} onKeyUp={this.handleKeyPress}>
+        <Container
+          isOpen={isOpen}
+          {...props}
+          align={align}
+          onClick={e => (disabled ? noop() : this.toggle(e))}
```
Comment (stereobooster) [inline src/ContextMenu/ContextMenu.tsx]:
```suggestion
          onClick={e => { if (!disabled) { this.toggle(e) } } }
```

File: src/ContextMenu/ContextMenu.tsx [hunk 6]
```diff
@@ -151,29 +159,56 @@ class ContextMenu extends React.Component<ContextMenuProps, Readonly<State>> {
...
+    } = this.props
     const isOpen = open || this.state.isOpen
     const renderedChildren = isChildAFunction(children) ? children(this.state.isOpen) : children
     return (
       <>
         {isOpen && <InvisibleOverlay onClick={this.toggle} />}
-        <Container isOpen={isOpen} {...props} align={align} onClick={this.toggle} onKeyUp={this.handleKeyPress}>
+        <Container
+          isOpen={isOpen}
+          {...props}
```
Comment (stereobooster) [inline src/ContextMenu/ContextMenu.tsx]: move `{...props}` above `isOpen={isOpen}`?

File: src/Select/Select.util.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+export const isOptionSelected = (value: any) => (option: IOption) => {
+  if (!Array.isArray(value)) {
+    return value === option.value
+  }
+
+  return value.includes(option.value)
+}
+
+export const getNewValue = (value: any) => (newValue: Value) => {
+  if (Array.isArray(value) && !value.includes(newValue)) {
```
Comment (stereobooster) [inline src/Select/Select.util.ts]:
```ts
if (Array.isArray(value)) {
  const newValueIndex = value.indexOf(newValue)
  if (newValueIndex === -1) {
    return [...value, newValue]
  } else {
    return [...value.slice(0, newValueIndex), ...value.slice(newValueIndex + 1)]
  }
  return newValue
}
```
?

File: src/Select/Select.types.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+export interface SelectPropsWithCustomOption extends BaseSelectProps {
+  /** Custom Option */
+  customOption?: string
+  /** Current value */
+  value: null | Value
+}
+
+export interface SelectPropsWithMultiSelect extends BaseSelectProps {
+  /** Custom Option */
+  customOption?: never
```
Comment (stereobooster) [inline src/Select/Select.types.ts]: DhMKJVoUwAAZWmw

File: src/Select/Select.tsx [hunk 9]
```diff
@@ -1,302 +1,134 @@
...
+  ...rest
+}) => {
+  const uniqueId = useUniqueId(id)
+  const [filter, setFilter] = React.useState("")
+  const [customInputValue, setCustomInputValue] = React.useState("")
+
+  const appendCustomOption = React.useCallback(
+    (options: IContextMenuItem[]): IContextMenuItem[] => {
+      // We can't have a multiselect _and_ a custom option.
+      if (Array.isArray(value) && process.env.NODE_ENV === "production") {
```
Comment (stereobooster) [inline src/Select/Select.tsx]:
```suggestion
      if (Array.isArray(value) && process.env.NODE_ENV !== "production") {
```

File: src/Select/Select.tsx [hunk 10]
```diff
@@ -1,302 +1,134 @@
...
+        },
+      })(options)
+    },
+    [customOption, customInputValue, value],
+  )
+
+  const filterComponent = React.useMemo(
+    () => (
+      <FilterInput
+        onClick={e => e.stopPropagation()}
```
Comment (stereobooster) [inline src/Select/Select.tsx]: not that this important, but it can be const, like noop. `const nooPropagation = e => e.stopPropagation()` 🤣

File: src/Select/Select.tsx [hunk 11]
```diff
@@ -1,302 +1,134 @@
...
-        {selectWithoutLabel}
-      </Label>
-    ) : (
-      selectWithoutLabel
+      label: option.label ? option.label : "",
+      isActive: isOptionSelected(value)(option),
+    }))(truncatedOptions)
+
+    return ([
+      [
```
Comment (stereobooster) [inline src/Select/Select.tsx]: 2eeunw

File: src/ContextMenu/ContextMenu.tsx (hunk 1 above)
Comment (TejasQ) [reply src/ContextMenu/ContextMenu.tsx]: auto import 🤦‍♂

File: src/Select/Select.util.ts [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+  if (Array.isArray(value) && value.includes(newValue)) {
+    const newValueIndex = value.indexOf(newValue)
+    return [...value.slice(0, newValueIndex), ...value.slice(newValueIndex + 1)]
+  }
+  return newValue
+}
+
+export const customInputSymbol = Symbol("custom input")
+
+export const getDisplayValue = (internalValue: any, customInputValue: any, customInputSymbol: symbol) => {
```
Comment (stereobooster) [inline src/Select/Select.util.ts]: `any`

**Comment by stereobooster** [comment]
5798d6b1dd08951e578b47e6-750-361

**Comment by stereobooster** [comment]
we need `preventDefault` for arrows

arrow-bug

File: src/ContextMenu/ContextMenu.Item.tsx (hunk 3 above)
Comment (TejasQ) [reply src/ContextMenu/ContextMenu.Item.tsx]: https://github.com/contiamo/operational-ui/issues/1006

File: src/ContextMenu/ContextMenu.Item.tsx (hunk 4 above)
Comment (TejasQ) [reply src/ContextMenu/ContextMenu.Item.tsx]: This constant is declared further down as the actual component. This is just a styled visual container.

**Comment by stereobooster** [comment]
- down/up arrow doesn't open dropdown
- after selecting item in the dropdown focus is lost. If I would want to change my selection need to type Tab, Shift + Tab, Enter.

**Comment by mpotomin** [comment]
Open a dropdown -> navigate within it with keyboard arrows -> Tab Tab Tab to find another Select on this screen -> Enter to open the second one:
Screenshot 2019-06-05 at 16 20 44

**Comment by TejasQ** [comment]
> - down/up arrow doesn't open dropdown

Fixed.

File: src/Select/Select.tsx [hunk 13]
```diff
@@ -1,302 +1,158 @@
...
+        >
+          {label && <LabelText id={`operational-ui__Select-Label-${uniqueId}`}>{label}</LabelText>}
+          <Combobox naked={Boolean(naked)}>
+            <SelectInput
+              disabled={disabled}
+              placeholder={placeholder}
+              readOnly={value !== customInputValue}
+              value={getDisplayValue(value, customInputValue, customInputSymbol)}
+              id={`operational-ui__Select-Input-${uniqueId}`}
+              onClick={e => {
```
Comment (mpotomin) [inline src/Select/Select.tsx]: Property 'onClick' does not exist on type '(IntrinsicAttributes & BasePropsWithCopy & .....

**Comment by mpotomin** [comment]
Let's add a drop shadow for the Context Menu, otherwise it looks barely distinguishable from the Table lines

Screenshot 2019-06-05 at 17 33 18

**Comment by TejasQ** [comment]
Fixed all open criticisms.

File: src/ContextMenu/ContextMenu.tsx [hunk 14]
```diff
@@ -82,11 +89,13 @@ const InvisibleOverlay = styled("div")(({ theme }) => ({
 
 class ContextMenu extends React.Component<ContextMenuProps, Readonly<State>> {
   private menu: HTMLDivElement | null = null
+  private uniqueId = this.props.id || (nanoid()[0] as string)
```
Comment (stereobooster) [inline src/ContextMenu/ContextMenu.tsx]: `nanoid()` returns string, `nanoid()[0]` is the first char of it

**Comment by mpotomin** [comment]
Would be nice to be able to style it. Currently if I apply

```typescript
const StyledSelect = styled(Select)`
  width: 50px;
`;
```
only the width of the ContextMenu is changing
Screenshot 2019-06-06 at 10 33 12

File: src/Select/Select.styled.ts [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+  display: grid;
+  grid-template-columns: auto 40px;
+  grid-gap: 1px;
+  align-items: stretch;
+  border: 1px solid ${({ theme }) => theme.color.border.select};
+  border-width: ${({ naked }) => (naked ? 0 : 1)}px;
+  border-radius: ${({ theme }) => theme.borderRadius}px;
+
+  :focus {
+    box-shadow: 0 0 0 1px ${({ theme }) => theme.color.primary};
```
Comment (mpotomin) [inline src/Select/Select.styled.ts]: We should make util out of the focused style to use it across the codebase

File: cypress/integration/Select.ts [hunk 16]
```diff
@@ -0,0 +1,156 @@
+describe("Select Components", () => {
+  beforeEach(() => {
+    cy.visit("/#!/Select")
+  })
+  it("should be visible in the DOM", () => {
+    cy.get('[data-cy="basic-select"]')
+  })
+  it("should be expandable via keyboard", () => {
+    Cypress.on("uncaught:exception", () => {
```
Comment (stereobooster) [inline cypress/integration/Select.ts]: I wonder why. Maybe we need to add small comment here

File: cypress/integration/Select.ts [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+      return false
+    })
+    cy.tab()
```
Comment (stereobooster) [inline cypress/integration/Select.ts]: this test is hard to follow, some meaningful name for `cy.tab().tab()...` would help

File: cypress/integration/Select.ts [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+      .focused()
+      .type("{downarrow}{enter}")
+    cy.get("body")
+      .tab()
+      .tab()
+      .tab()
+      .type("{enter}")
+      .focused()
+      .type("{downarrow}{end}{enter}")
+    cy.get('[data-cy="custom-select"] input[value="chickenshola amigo como estas"]')
```
Comment (stereobooster) [inline cypress/integration/Select.ts]: I have no idea what is happening here

File: src/ContextMenu/ContextMenu.Item.tsx [hunk 19]
```diff
@@ -131,14 +147,16 @@ const ContextMenuItemIcon: React.SFC<Pick<Props, "item" | "iconLocation">> = pro
...
-    {(!props.iconLocation || props.iconLocation === "left") && (
-      <ContextMenuItemIcon iconLocation={props.iconLocation} item={props.item} />
-    )}
-    <Content value={props.item} />
-    {props.iconLocation === "right" && <ContextMenuItemIcon iconLocation={props.iconLocation} item={props.item} />}
-  </Container>
-)
+const ContextMenuItem: React.SFC<Props> = props => {
+  return (
+    <Container {...props} condensed={props.condensed}>
```
Comment (stereobooster) [inline src/ContextMenu/ContextMenu.Item.tsx]:
```suggestion
    <Container {...props}>
```

File: src/ContextMenu/ContextMenu.tsx [hunk 20]
```diff
@@ -131,49 +185,79 @@ class ContextMenu extends React.Component<ContextMenuProps, Readonly<State>> {
...
+        <Container
+          {...props}
+          aria-activedescendant={
+            isOpen ? `operational-ui__ContextMenuItem-${this.uniqueId}-${this.state.focusedItemIndex}` : undefined
+          }
+          aria-disabled={Boolean(disabled)}
+          aria-expanded={isOpen}
+          isOpen={isOpen}
+          align={align}
+          onClick={() => {
```
Comment (stereobooster) [inline src/ContextMenu/ContextMenu.tsx]: `onClick={ disabled ? undefined : this.toggle }` optional

File: src/ContextMenu/ContextMenu.tsx [hunk 21]
```diff
@@ -186,6 +270,13 @@ class ContextMenu extends React.Component<ContextMenuProps, Readonly<State>> {
             </MenuContainer>
           )}
         </Container>
+        {/* Element to close an open select when blurring it so only one can be open at a time */}
+        <div
+          tabIndex={this.state.isOpen ? 0 : -1}
+          role="button"
+          onFocus={() => this.setState(() => ({ isOpen: false }))}
```
Comment (stereobooster) [inline src/ContextMenu/ContextMenu.tsx]: comment says "when blurring",  but this has `onFocus`

**Comment by TejasQ** [comment]
Merging since everything's green and no known issues (and approved) except https://github.com/contiamo/operational-ui/issues/1016
